### PR TITLE
test(contracts): expand Soroban coverage for strategy and access control

### LINKED ISSUE #53: Write unit tests for allocation strategy contract (#19)
## Overview

Test companion for #19 — unit tests for the allocation strategy contract that determines how vault funds are distributed across yield sources.

### Strategy Initialization
- Strategy initializes with correct vault type (Conservative, Balanced, Growth, DeFi500)
- Each vault type has a defined default allocation table
- Re-initialization is rejected

### Allocation Computation
- `compute_allocation(total_amount, apys)` returns weights that sum to exactly 100%
- Conservative strategy caps individual protocol weight at the defined maximum
- Growth strategy allocates more weight to higher-APY sources
- DeFi500 strategy produces an even distribution across registered sources
- Zero APY source receives zero allocation weight

### Rebalancing Threshold
- `needs_rebalance(current_weights, target_weights)` returns `true` when drift exceeds threshold
- Returns `false` when weights are within tolerance
- Threshold value is configurable by admin

### Constraint Enforcement
- Allocation to a single source never exceeds the per-strategy maximum weight
- Allocation to a deactivated/unregistered source is zero

### Admin Controls
- Admin can update strategy parameters (thresholds, max weights)
- Non-admin update attempts are rejected

## Test Type

Soroban unit tests. Math-heavy — test with precise i128 values to catch rounding errors.

## Acceptance Criteria

- [ ] Allocation weights sum invariant is tested across all strategies
- [ ] Rebalancing threshold boundary conditions are tested
- [ ] Tests pass with `cargo test`

## Dependencies

- Closes after #19 is merged

### LINKED ISSUE #54: Write unit tests for admin access control module (#20)
## Overview

Test companion for #20 — unit tests for the cross-contract access control module covering role assignment, permission checks, and the emergency pause mechanism.

### Role Management
- Admin can grant a role to an address
- Admin can revoke a role from an address
- Granting a role that is already held is idempotent (no error)
- Revoking a role that was never held is idempotent
- Non-admin attempting to grant or revoke is rejected

### Permission Checks
- `has_role(address, role)` returns `true` when address holds the role
- `has_role(address, role)` returns `false` when role is absent
- `require_role(address, role)` panics/returns error when role is absent
- `require_role(address, role)` succeeds silently when role is present

### Emergency Pause
- Admin can pause the protocol — `is_paused()` returns `true`
- Pausing an already-paused protocol is idempotent
- Admin can unpause — `is_paused()` returns `false`
- Non-admin attempting to pause is rejected
- Contracts that call `require_not_paused()` are blocked when paused

### Multi-Contract Scenarios
- Role check works when called from a different contract (cross-contract auth)
- Pause state is visible to contracts that query it

## Test Type

Soroban unit tests. Cross-contract scenarios using multiple contract instances in the test environment.

## Acceptance Criteria

- [ ] All role operations (grant, revoke, check) are covered
- [ ] Pause/unpause state transitions are fully tested
- [ ] Cross-contract auth scenario has at least one test
- [ ] Tests pass with `cargo test`

## Dependencies

- Closes after #20 is merged

### PR DESCRIPTION
**Pull request by josephchimebuka**
## Summary
- Expand access control coverage with explicit grant idempotence naming, require_role error-path coverage, and revoke idempotence checks.
- Expand vault pause coverage with stronger idempotence assertions plus split cross-contract pause success and failure tests.
- Extend allocation strategy coverage to include vault-type initialization defaults, APY-driven allocation computation, zero-weight handling for inactive and unregistered sources, admin-updatable rebalance parameters, and threshold boundary checks.
- Add the missing allocation strategy helpers required to exercise the issue acceptance criteria: strategy presets, APY-based weight computation, and rebalance detection.

## Testing
- `cargo test` under `packages/contracts`

Closes #53
Closes #54

Related: #19, #20

### CODE REVIEW

File: packages/contracts/contracts/access_control/src/test.rs [hunk 1]
```diff
@@ -222,6 +222,21 @@ fn admin_can_revoke_operator_role() {
...
+    let (env, admin, other, cid) = setup();
+
+    invoke(&env, &cid, || {
+        AccessControl::revoke_role(&env, &admin, &other, Role::Operator)
+    });
+
+    assert!(!read(&env, &cid, || AccessControl::has_role(
+        &env,
+        &other,
+        Role::Operator
```
Comment (0xDeon) [inline packages/contracts/contracts/access_control/src/test.rs]: **Missing: grant idempotence test (Issue #54 requirement)**

Issue #54 explicitly requires:
> *Granting a role that is already held is idempotent (no error)*

You've covered revoke idempotence here but the grant counterpart is missing. Add a `granting_existing_role_is_idempotent` test that grants the same role twice and asserts no error + `has_role` still returns `true`.

Also missing from #54: a `require_role` error path test — `require_role(address, role)` should panic/error when the role is absent.
